### PR TITLE
[VPlan] Compute scalable VF in preheader for induction increment.

### PR DESCRIPTION
**Pull request by fhahn**
UF * VF is loop invariant and can be computed directly in the preheader.
This prepares the code for #74333 and reduces the test changes.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 1]
```diff
@@ -340,8 +340,13 @@ Value *VPInstruction::generateInstruction(VPTransformState &State,
       auto *Phi = State.get(getOperand(0), 0);
       // The loop step is equal to the vectorization factor (num of SIMD
       // elements) times the unroll factor (num of SIMD instructions).
-      Value *Step =
-          createStepForVF(Builder, Phi->getType(), State.VF, State.UF);
+      Value *Step;
+      {
+        BasicBlock *VectorPH = State.CFG.getPreheaderBBFor(this);
+        IRBuilder<>::InsertPointGuard Guard(Builder);
+        Builder.SetInsertPoint(VectorPH->getTerminator());
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:347]: This is fine. Can alternatively define a separate builder set to the vector preheader to serve createStepForVF, independent of `Builder` which serves CreateAdd.

Perhaps worth a comment that step for VF must be loop invariant, and for scalable VF's generates a call to vscale which should be placed in the preheader.

**Review by ayalz — APPROVED** [review]
Thanks for breaking this piecemeal!
Adding a minor nit.

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 1 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: Adjusted in the latest version, thanks!

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 2]
```diff
@@ -340,8 +340,14 @@ Value *VPInstruction::generateInstruction(VPTransformState &State,
...
-      Value *Step =
-          createStepForVF(Builder, Phi->getType(), State.VF, State.UF);
+      Value *Step;
+      {
+        BasicBlock *VectorPH = State.CFG.getPreheaderBBFor(this);
+        IRBuilder<> PHBuilder(VectorPH->getTerminator());
+        // Step is loop-invariant, calls to vscale will be placed in the
+        // preheader.
+        Step = createStepForVF(PHBuilder, Phi->getType(), State.VF, State.UF);
+      }
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:350]:
```suggestion
      // Step is loop-invariant, calls to vscale will be placed in the preheader.
      BasicBlock *VectorPH = State.CFG.getPreheaderBBFor(this);
      IRBuilder<> PHBuilder(VectorPH->getTerminator());
      Value *Step = createStepForVF(PHBuilder, Phi->getType(), State.VF, State.UF);
```

(no need for the bracketed block, which was needed for the point guard.)

**Review by ayalz — COMMENTED** [review]
post-commit nit

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 2 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:350]: The cow has been remove in https://github.com/llvm/llvm-project/commit/a5891fa4d2b76cf9dec96da9ded59fc4937d3342

**Pull request merged**